src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/DefaultRazorSemanticTokenInfoService.cs

```diff
@@ -0,0 +1,213 @@
+            foreach (var token in syntaxTokens)
+            {
+                var newData = GetData(token, previousToken, razorCodeDocument, semanticTokensLegend);
+                data.AddRange(newData);
+
+                previousToken = token;
+            }
+
+            return new SemanticTokens
+            {
+                Data = data
+            };
+        }
+
+        /**
+         * In short, each token takes 5 integers to represent, so a specific token `i` in the file consists of the following array indices:
+         *  - at index `5*i`   - `deltaLine`: token line number, relative to the previous token
+         *  - at index `5*i+1` - `deltaStart`: token start character, relative to the previous token (relative to 0 or the previous token's start if they are on the same line)
+         *  - at index `5*i+2` - `length`: the length of the token. A token cannot be multiline.
+         *  - at index `5*i+3` - `tokenType`: will be looked up in `SemanticTokensLegend.tokenTypes`
+         *  - at index `5*i+4` - `tokenModifiers`: each set bit will be looked up in `SemanticTokensLegend.tokenModifiers`
+        **/
+        private static IEnumerable<uint> GetData(
+            SyntaxNode currentNode,
+            SyntaxNode previousNode,
+            RazorCodeDocument razorCodeDocument,
+            SemanticTokenLegend legend)
+        {
+            var previousRange = previousNode?.GetRange(razorCodeDocument);
+            var currentRange = currentNode.GetRange(razorCodeDocument);
+
+            // deltaLine
+            var previousLineIndex = previousNode == null ? 0 : previousRange.Start.Line;
+            yield return (uint)(currentRange.Start.Line - previousLineIndex);
+
+            // deltaStart
+            if (previousRange != null && previousRange?.Start.Line == currentRange.Start.Line)
+            {
+                yield return (uint)(currentRange.Start.Character - previousRange.Start.Character);
+            }
+            else
+            {
+                yield return (uint)(currentRange.Start.Character);
+            }
+
+            // length
+            if(currentNode.Span.Length <= 0)
+            {
+                throw new NotImplementedException();
+            }
+            yield return (uint)currentNode.Span.Length;
+
+            // tokenType]
+            yield return (uint)GetTokenTypeData(currentNode, legend);
+
+            // tokenModifiers
+            yield return (uint)GetTokenModifierData(currentNode, legend);
+        }
+
+        private static long GetTokenTypeData(SyntaxNode syntaxToken, SemanticTokenLegend legend)
+        {
+            switch(syntaxToken.Parent.Kind)
+            {
+                case SyntaxKind.MarkupTagHelperStartTag:
+                    return legend.TokenTypesLegend["razorTagHelperElementStartTag"];
+                case SyntaxKind.MarkupTagHelperEndTag:
+                    return legend.TokenTypesLegend["razorTagHelperElementEndTag"];
```

Comment:
For the legend I'm not confident we need to make a distinction between start/end tags (also below minimized/unminimized/directive  attributes). I think if we had a single token type for start/end tags and a single token type for attributes that'd be sufficient. Too much flexibility will make configuring these things a pain for users.

Reply:
I had originally made these separate with the understanding that they would be translated into textmate scopes and users could just do `razor.taghelper.element.*`, but still have the option to do specific stuff for end-tags if it so pleased them. Given that it seems the model is NOT to translate semantic tokens into textmate scopes I agree with you, I'll pare these down into just important bits.

Reply:
Ahhhh i see, ya that would have made perfect sense.